### PR TITLE
Use iss claim as principal for access token, added validation for all…

### PR DESCRIPTION
**Pull request by sanjeevchopra**
…owed_device_id. plus some cleanup.

### CODE REVIEW

File: server/src/test/java/org/cloudfoundry/identity/uaa/provider/token/JwtBearerAssertionTokenAuthenticatorTest.java [hunk 1]
```diff
@@ -136,37 +171,37 @@ public String getPublicKey(final String tenantId, final String deviceId) throws
...
-        String token = new MockAssertionToken().mockAssertionToken(DEVICE_1_ID, System.currentTimeMillis() - 240000,
-                600, TENANT_ID, AUDIENCE);
+        String token = new MockAssertionToken().mockAssertionToken(DEVICE_1_CLIENT_ID, DEVICE_1_ID,
+                System.currentTimeMillis(), 600, TENANT_ID, AUDIENCE);
         
         //remove d1 client from mock
-        when(this.clientDetailsService.loadClientByClientId(DEVICE_1_ID))
+        when(this.clientDetailsService.loadClientByClientId(DEVICE_1_CLIENT_ID))
         .thenReturn(null);
         this.tokenAuthenticator.setClientDetailsService(this.clientDetailsService);
```
Comment (irinaepshteyn) [inline server/src/test/java/org/cloudfoundry/identity/uaa/provider/token/JwtBearerAssertionTokenAuthenticatorTest.java]: loadClientByClientId could throw a NoSuchClient and InvalidClient exceptions or return an empty string.  Should we test that here?

File: server/src/test/java/org/cloudfoundry/identity/uaa/provider/token/JwtBearerAssertionTokenAuthenticatorTest.java [hunk 2]
```diff
@@ -176,43 +211,43 @@ public void testTokenWithNoExpClaim() {
         String token = testTokenUtil.mockInvalidExpirationAssertionToken(DEVICE_1_ID, 0, TENANT_ID,
                 AUDIENCE, null);
         this.tokenAuthenticator.setClientDetailsService(this.clientDetailsService);
-        this.tokenAuthenticator.authenticate(token);
+        this.tokenAuthenticator.authenticateWithoutClientAssertionHeader(token);
     }
```
Comment (irinaepshteyn) [inline server/src/test/java/org/cloudfoundry/identity/uaa/provider/token/JwtBearerAssertionTokenAuthenticatorTest.java]: Should we test for an assertion token with missing (null or empty) subject and issuer claims as well?

File: server/src/test/java/org/cloudfoundry/identity/uaa/provider/token/JwtBearerAssertionTokenAuthenticatorTest.java [hunk 3]
```diff
@@ -37,38 +39,70 @@
...
+        long currentTime = System.currentTimeMillis();
+        String token = new MockAssertionToken().mockAssertionToken(DEVICE_1_CLIENT_ID, DEVICE_1_ID,
+                currentTime, 600, TENANT_ID, AUDIENCE);
+        String header = new MockClientAssertionHeader().mockSignedHeader(this.currentTimeSecs, DEVICE_1_ID, TENANT_ID);
+        
+        //allowed_device_id not set
+        when(this.clientDetailsService.loadClientByClientId(DEVICE_1_CLIENT_ID)) 
+            .thenReturn(new BaseClientDetails(DEVICE_1_CLIENT_ID, null, null, null, null, null));
+        this.tokenAuthenticator.setClientDetailsService(this.clientDetailsService);
+
```
Comment (irinaepshteyn) [inline server/src/test/java/org/cloudfoundry/identity/uaa/provider/token/JwtBearerAssertionTokenAuthenticatorTest.java]: We should test for allowed device being empty as well.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/JwtBearerGrantIT.java [hunk 4]
```diff
@@ -98,11 +102,11 @@ public void testJwtBearerGrantForUnknownClient() {
```
Comment (irinaepshteyn) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/JwtBearerGrantIT.java]: Should we add an integration test for the existent client not authorized for the device?  Similar to testNoAllowedDeviceIdSet unit test...

**Comment by sanjeevchopra** [comment]
Thanks for the review. Tests you point out would be nice to have.

File: server/src/test/java/org/cloudfoundry/identity/uaa/provider/token/JwtBearerAssertionTokenAuthenticatorTest.java (hunk 3 above)
Comment (sanjeevchopra) [inline server/src/test/java/org/cloudfoundry/identity/uaa/provider/token/JwtBearerAssertionTokenAuthenticatorTest.java]: valid point, yet it seems a bit excessive. we should have a common standard on how to test json input like this.

File: uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/JwtBearerGrantIT.java (hunk 4 above)
Comment (sanjeevchopra) [inline uaa/src/test/java/org/cloudfoundry/identity/uaa/integration/feature/JwtBearerGrantIT.java]: not sure - this is another case worth discussing. We generally try to minimize integration tests where something can be covered in unit tests. On other hand, for important use cases like this, a integration test could ensure that involved components are wired correctly?

File: server/src/test/java/org/cloudfoundry/identity/uaa/provider/token/JwtBearerAssertionTokenAuthenticatorTest.java (hunk 1 above)
Comment (sanjeevchopra) [inline server/src/test/java/org/cloudfoundry/identity/uaa/provider/token/JwtBearerAssertionTokenAuthenticatorTest.java]: seems worthwhile in a separate unit test.   this method does declare a runtime exception on the interface.

File: server/src/test/java/org/cloudfoundry/identity/uaa/provider/token/JwtBearerAssertionTokenAuthenticatorTest.java (hunk 3 above)
Comment (irinaepshteyn) [inline server/src/test/java/org/cloudfoundry/identity/uaa/provider/token/JwtBearerAssertionTokenAuthenticatorTest.java]: I agree that we need to have json validation even before we proceed with authenticating jwt assertion.  During validation we should check that header and assertion contain all the expected claims and claims are not null and not empty.  
In this case we are validating uaa client attribute.  We test if expected attribute is not set and if it's set but not authorized for the device.  We should also test it expected attribute is set to null or empty.

File: server/src/main/java/org/cloudfoundry/identity/uaa/provider/token/JwtBearerAssertionTokenAuthenticator.java [hunk 5]
```diff
@@ -121,14 +131,58 @@ private Authentication authenticateJwtAssertionToken(final Jwt jwt, String devic
...
+        }
+
+        // require tenant_id claim
+        if (!StringUtils.hasText((String) claims.get(ClaimConstants.TENANT_ID))) {
+            throw new InvalidTokenException("Tenant id claim is required in jwt-bearer assertion.");
+        }
+
+        // require issuer claim
+        if (!StringUtils.hasText((String) claims.get(ClaimConstants.ISS))) {
+            throw new InvalidTokenException("Issuer claim is required in jwt-bearer assertion.");
```
Comment (sanjeevchopra) [inline server/src/main/java/org/cloudfoundry/identity/uaa/provider/token/JwtBearerAssertionTokenAuthenticator.java]: for validation errors, pls use the actual claim names in the errors. 'iss' in this case.
